File: packages/zako3-data/src/schemas/playback.ts
import { z } from 'zod';

export const audioMetadataSchema = z.object({
    type: z.string(),
    value: z.string(),
});

export const trackSchema = z.object({
    trackId: z.string(),
    queueName: z.string(),
    metadata: z.array(audioMetadataSchema),
    tapName: z.string(),
    audioRequestString: z.string(),
    requestedBy: z.string(),
    volume: z.number(),
    paused: z.boolean(),
});

export const discordUserInfoSchema = z.object({
    id: z.string(),
    name: z.string(),
    avatarUrl: z.string().nullable().optional(),
});

export const queueMetaSchema = z.object({
    user: discordUserInfoSchema.optional(),
});

export const guildPlaybackStateSchema = z.object({
    guildId: z.string(),
    guildName: z.string().default(''),
    guildIconUrl: z.string().optional(),
    channelId: z.string(),
    channelName: z.string().default(''),
    queues: z.record(z.string(), z.array(trackSchema)),
    queueMeta: z.record(z.string(), queueMetaSchema).default({}),
});

export const playbackActionSchema = z.object({
    id: z.string(),
    actionType: z.string(),
    guildId: z.string(),
    channelId: z.string(),
    actorDiscordUserId: z.string(),
    createdAt: z.string().datetime(),
    undoneAt: z.string().datetime().nullable(),
    undoneByDiscordUserId: z.string().nullable(),
});

export const stopTrackSchema = z.object({
    guildId: z.string(),
    channelId: z.string(),
    trackId: z.string(),
});

export const skipSchema = z.object({
    guildId: z.string(),
    channelId: z.string(),
});

export const queueOperationSchema = z.object({
    op: z.enum(['remove', 'set_volume']),
    trackId: z.string(),
    targetIndex: z.number().optional(),
    volume: z.number().optional(),
});

export const editQueueSchema = z.object({
    guildId: z.string(),
    channelId: z.string(),
    operations: z.array(queueOperationSchema),
});

export const pauseTrackSchema = z.object({
    guildId: z.string(),
    channelId: z.string(),
    trackId: z.string(),
});

export const resumeTrackSchema = z.object({
    guildId: z.string(),
    channelId: z.string(),
    trackId: z.string(),
});

export type AudioMetadataDto = z.infer<typeof audioMetadataSchema>;
export type TrackDto = z.infer<typeof trackSchema>;
export type DiscordUserInfoDto = z.infer<typeof discordUserInfoSchema>;
export type QueueMetaDto = z.infer<typeof queueMetaSchema>;
export type GuildPlaybackStateDto = z.infer<typeof guildPlaybackStateSchema>;
export type PlaybackActionDto = z.infer<typeof playbackActionSchema>;
export type StopTrackDto = z.infer<typeof stopTrackSchema>;
export type SkipDto = z.infer<typeof skipSchema>;
export type QueueOperationDto = z.infer<typeof queueOperationSchema>;
export type EditQueueDto = z.infer<typeof editQueueSchema>;
export type PauseTrackDto = z.infer<typeof pauseTrackSchema>;
export type ResumeTrackDto = z.infer<typeof resumeTrackSchema>;

File: types/src/hq/playback.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use utoipa::ToSchema;

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct AudioMetadataDto {
    pub r#type: String,
    pub value: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct TrackDto {
    pub track_id: String,
    pub queue_name: String,
    pub metadata: Vec<AudioMetadataDto>,
    pub tap_name: String,
    pub audio_request_string: String,
    pub requested_by: String,
    pub volume: f32,
    pub paused: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct GuildPlaybackStateDto {
    pub guild_id: String,
    pub guild_name: String,
    pub guild_icon_url: Option<String>,
    pub channel_id: String,
    pub channel_name: String,
    pub queues: HashMap<String, Vec<TrackDto>>,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct PlaybackActionDto {
    pub id: String,
    pub action_type: String,
    pub guild_id: String,
    pub channel_id: String,
    pub actor_discord_user_id: String,
    pub created_at: DateTime<Utc>,
    pub undone_at: Option<DateTime<Utc>>,
    pub undone_by_discord_user_id: Option<String>,
}

#[derive(Debug, Clone, Deserialize, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct StopTrackDto {
    pub guild_id: String,
    pub channel_id: String,
    pub track_id: String,
}

#[derive(Debug, Clone, Deserialize, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct SkipDto {
    pub guild_id: String,
    pub channel_id: String,
}

#[derive(Debug, Clone, Deserialize, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct QueueOperation {
    /// "remove" or "set_volume"
    pub op: String,
    pub track_id: String,
    pub target_index: Option<usize>,
    pub volume: Option<f32>,
}

#[derive(Debug, Clone, Deserialize, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct EditQueueDto {
    pub guild_id: String,
    pub channel_id: String,
    pub operations: Vec<QueueOperation>,
}

#[derive(Debug, Clone, Deserialize, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct PauseTrackDto {
    pub guild_id: String,
    pub channel_id: String,
    pub track_id: String,
}

#[derive(Debug, Clone, Deserialize, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct ResumeTrackDto {
    pub guild_id: String,
    pub channel_id: String,
    pub track_id: String,
}
